File: ios/Sources/Omnesis/UI/Agent/AgentTranscriptSkeleton.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

/// Placeholder transcript shown while a resumed conversation's messages are
/// still loading. Opening a stored conversation switches the surface to the
/// target the instant it's tapped (`AgentCoordinator.beginResume`); this
/// skeleton fills the brief gap until the real transcript lands, so
/// navigation never blocks on the network and the user always sees *a*
/// conversation taking shape rather than the previous one lingering.
///
/// A few shimmering bubble placeholders — alternating assistant (leading,
/// multi-line) and user (trailing pill) — approximate a real transcript's
/// rhythm so the moment reads as "this conversation is loading", not "empty".
@available(iOS 17.0, *)
struct AgentTranscriptSkeleton: View {
    /// Drives the highlight sweep. Animated on appear; a static mid-sweep
    /// value is perfectly legible in snapshot renders.
    @State private var shimmerPhase: CGFloat = -0.6

    var body: some View {
        GeometryReader { geo in
            let width = max(geo.size.width - Theme.Spacing.md * 2, 0)
            VStack(alignment: .leading, spacing: 16) {
                assistantBlock(width: width, lineFractions: [0.86, 0.72, 0.5])
                userBlock(width: width, fraction: 0.55)
                assistantBlock(width: width, lineFractions: [0.9, 0.76])
                userBlock(width: width, fraction: 0.4)
                assistantBlock(width: width, lineFractions: [0.8, 0.64, 0.68, 0.36])
            }
            .padding(.horizontal, Theme.Spacing.md)
            // Clear the floating menu / new-conversation buttons at the top.
            .padding(.top, 72)
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        }
        .allowsHitTesting(false)
        .accessibilityElement()
        .accessibilityLabel("Loading conversation")
        .onAppear {
            withAnimation(.linear(duration: 1.25).repeatForever(autoreverses: false)) {
                shimmerPhase = 1.3
            }
        }
    }

    /// A leading-aligned stack of placeholder lines — an assistant reply.
    private func assistantBlock(width: CGFloat, lineFractions: [CGFloat]) -> some View {
        VStack(alignment: .leading, spacing: 9) {
            ForEach(Array(lineFractions.enumerated()), id: \.offset) { _, fraction in
                bar(width: width * fraction)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// A trailing-aligned accent pill — a user message.
    private func userBlock(width: CGFloat, fraction: CGFloat) -> some View {
        HStack(spacing: 0) {
            Spacer(minLength: 0)
            bar(
                width: width * fraction,
                height: 34,
                tint: Theme.accent.opacity(0.16),
                // Match the real user bubble's corner radius (AgentBubbleViews)
                // so the placeholder reads as a message, not a boxy block.
                cornerRadius: 12
            )
        }
    }

    /// One shimmering placeholder bar. The base fill keeps the shape visible
    /// at every animation phase (and in a static snapshot); a soft highlight
    /// strip sweeps across it left-to-right.
    private func bar(
        width: CGFloat,
        height: CGFloat = 13,
        tint: Color = Theme.bgSecondary,
        cornerRadius: CGFloat = Theme.Radius.medium
    )
        -> some View {
        let shape = RoundedRectangle(cornerRadius: cornerRadius)
        return shape
            .fill(tint)
            .frame(width: width, height: height)
            .overlay(
                LinearGradient(
                    colors: [.clear, Color.white.opacity(0.22), .clear],
                    startPoint: .leading,
                    endPoint: .trailing
                )
                .frame(width: max(width * 0.4, 1))
                .offset(x: shimmerPhase * width)
            )
            .clipShape(shape)
    }
}

#if DEBUG
@available(iOS 17.0, *)
#Preview("AgentTranscriptSkeleton — dark") {
    AgentTranscriptSkeleton()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Theme.bgPrimary)
        .environment(\.colorScheme, .dark)
}

@available(iOS 17.0, *)
#Preview("AgentTranscriptSkeleton — light") {
    AgentTranscriptSkeleton()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Theme.bgPrimary)
        .environment(\.colorScheme, .light)
}
#endif
#endif

File: android/app/src/main/kotlin/dev/omnesis/android/ui/agent/AgentTranscriptSkeleton.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.ui.agent

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalInspectionMode
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import dev.omnesis.android.designsystem.theme.OmTheme

/**
 * Placeholder transcript shown while a resumed conversation's messages are still loading.
 * Opening a stored conversation switches the surface to the target the instant it's tapped
 * ([AgentCoordinator.beginResume]); this skeleton fills the brief gap until the real transcript
 * lands, so navigation never blocks on the network and the user always sees *a* conversation
 * taking shape rather than the previous one lingering. Mirrors the iOS `AgentTranscriptSkeleton`.
 *
 * A few breathing placeholder bubbles — alternating assistant (leading, multi-line) and user
 * (trailing accent pill) — approximate a real transcript's rhythm. The breathing pulse is gated
 * on [LocalInspectionMode] so Roborazzi screenshot capture (which never idles an infinite
 * animation — see android/CLAUDE.md) freezes at a representative frame instead of hanging.
 */
@Composable
fun AgentTranscriptSkeleton(modifier: Modifier = Modifier) {
    val pulse = if (LocalInspectionMode.current) {
        0.55f
    } else {
        val transition = rememberInfiniteTransition(label = "skeletonPulse")
        transition.animateFloat(
            initialValue = 0.35f,
            targetValue = 0.75f,
            animationSpec = infiniteRepeatable(tween(900), RepeatMode.Reverse),
            label = "skeletonAlpha",
        ).value
    }
    Column(
        modifier
            .fillMaxSize()
            // Sit below the status bar, then clear the floating menu / new-conversation buttons
            // (which are themselves status-bar-inset in AgentScreen) — inset-aware so the first
            // bar never slides behind them on a tall notch.
            .windowInsetsPadding(WindowInsets.statusBars)
            .padding(horizontal = 12.dp)
            .padding(top = 56.dp)
            .semantics { contentDescription = "Loading conversation" },
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        AssistantBlock(listOf(0.86f, 0.72f, 0.50f), pulse)
        UserPill(0.55f, pulse)
        AssistantBlock(listOf(0.90f, 0.76f), pulse)
        UserPill(0.40f, pulse)
        AssistantBlock(listOf(0.80f, 0.64f, 0.68f, 0.36f), pulse)
    }
}

/** A leading-aligned stack of placeholder lines — an assistant reply. */
@Composable
private fun AssistantBlock(lineFractions: List<Float>, pulse: Float) {
    Column(verticalArrangement = Arrangement.spacedBy(9.dp)) {
        lineFractions.forEach { fraction -> Bar(fraction, pulse) }
    }
}

/** A trailing-aligned accent pill — a user message, matching the real bubble's corner radius. */
@Composable
private fun UserPill(fraction: Float, pulse: Float) {
    Row(Modifier.fillMaxWidth()) {
        Spacer(Modifier.weight(1f))
        Box(
            Modifier
                .fillMaxWidth(fraction)
                .height(34.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(OmTheme.colors.accent.copy(alpha = 0.10f + 0.12f * pulse)),
        )
    }
}

/** One breathing placeholder line. */
@Composable
private fun Bar(fraction: Float, pulse: Float) {
    Box(
        Modifier
            .fillMaxWidth(fraction)
            .height(13.dp)
            .clip(RoundedCornerShape(6.dp))
            .background(OmTheme.colors.bgTertiary.copy(alpha = pulse)),
    )
}
